### PR TITLE
Support convert orc timestamptz

### PR DESCRIPTION
**Pull request by zzzzming95**
### What changes were proposed in this pull request?

A user was attempting to convert an ORC backed external table in hive to a Iceberg table using the migrate command but was immediately met with a "Can not promote TIMESTAMP to TIMESTAMP" error. This occurs because our Spark -> Iceberg conversion code always converts to a Timestamp.withZone.

it relate issue : https://github.com/apache/iceberg/issues/9784

```
spark.sql("CREATE EXTERNAL TABLE mytable (foo timestamp) STORED AS orc LOCATION '/Users/russellspitzer/Temp/foo'") 
spark.sql("INSERT INTO mytable VALUES (now())")
 spark.sql("CALL spark_catalog.system.migrate('mytable')") 
 spark.sql("SELECT * FROM mytable")
```

```
java.lang.IllegalArgumentException: Can not promote TIMESTAMP type to TIMESTAMP
	at org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument(Preconditions.java:441)
	at org.apache.iceberg.orc.ORCSchemaUtil.buildOrcProjection(ORCSchemaUtil.java:301)
	at org.apache.iceberg.orc.ORCSchemaUtil.buildOrcProjection(ORCSchemaUtil.java:275)
	at org.apache.iceberg.orc.ORCSchemaUtil.buildOrcProjection(ORCSchemaUtil.java:258)
```
add a config , `iceberg.orc.convert.timestamptz` , when set is as true , it will auto convert orc `TIMESTAMP` as orc `TIMESTAMP_INSTANT` , so we can fix this issue.

### Why are the changes needed?
fix issue.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
add a UT

### CODE REVIEW

**Comment by tanvn** [comment]
@snazy @nastra @rdblue 
I confirmed that this issue is happening on my env (Spark 3.4, Iceberg 1.3.1) as well and this is blocking my team from migrating our Hive tables to Iceberg.

Could you please take a look at this PR? If there is anything I can do, please let me know (cc @zzzzming95 )

File: .palantir/revapi.yml [hunk 1]
```diff
@@ -1018,6 +1018,13 @@ acceptedBreaks:
       old: "method void org.apache.iceberg.PositionDeletesTable.PositionDeletesBatchScan::<init>(org.apache.iceberg.Table,\
         \ org.apache.iceberg.Schema, org.apache.iceberg.TableScanContext)"
       justification: "Removing deprecated code"
+    org.apache.iceberg:iceberg-orc:
```
Comment (nastra) [inline .palantir/revapi.yml:1021]: we can't break existing APIs but you could add an overloaded version of that method that takes an additional parameter. That way it won't break the existing API. See also https://iceberg.apache.org/contribute/#adding-new-functionality-without-breaking-apis

**Comment by nastra** [comment]
can you please rebase the PR and resolve conflicts? I'll try to take a closer look this week

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java [hunk 2]
```diff
@@ -23,7 +23,9 @@
 import java.util.Objects;
 import java.util.Optional;
 import java.util.stream.Collectors;
+import org.apache.hadoop.conf.Configuration;
```
Comment (snazy) [inline orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java:26]: Wonder whether the "whole `Configuration` spiel" is necessary here. Hadoop dependencies are already `compileOnly` to eventually get rid of those entirely. Wouldn't a simple boolean flag do the same thing in this class?

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java [hunk 3]
```diff
@@ -296,24 +304,36 @@ private static TypeDescription buildOrcProjection(
...
         if (mapping.containsKey(fieldId)) {
           TypeDescription originalType = mapping.get(fieldId).type();
           Optional<TypeDescription> promotedType = getPromotedType(type, originalType);
 
-          if (promotedType.isPresent()) {
+          boolean convertTimestampTZ =
+              config.getBoolean(ConfigProperties.ORC_CONVERT_TIMESTAMPTZ, false);
+
+          if (convertTimestampTZ
+              && type.typeId() == Type.TypeID.TIMESTAMP
```
Comment (snazy) [inline orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java:333]: Guess this deserves a new `case TIMESTAMP`, not an `if` here.

**Review by snazy — COMMENTED** [review]
Can't say much about the general approach wrt type mapping, but left some comments.

**Comment by tanvn** [comment]
@zzzzming95 
Do you have time and effort for this issue? (I would appreciate if you could)
If not, I might create a new PR based from this.

**Comment by tanvn** [comment]
@nastra @snazy 
I think your comments have been addressed by @zzzzming95 
Could you take another look please?

**Review by snazy — COMMENTED** [review]
On #9784 you mentioned:

> I think this is because hive and spark treat `timestamp` data type as timestamp with time zone and the orc file format is also stored as orc `timestamp` type. But in fact the hive `timestamp` data type should be stored as `timestamp_instant` in the orc file.

This sounds like Hive and Spark treat the `timestamp` type in a "less flexible way". IIRC Spark 3.4 introduced `timestamp_ntz`, whereas Hive uses no timezone at all.

The `iceberg.orc.convert.timestamptz` option introduced with this PR seems to be a global setting, so it affects all tables. I wonder whether this should rather be an ORC type property.

**Comment by pvary** [comment]
@deniskuzZ: How this is solved in Hive?

**Comment by findepi** [comment]
@raunaqmorarka please take a look

**Comment by deniskuzZ** [comment]
> @deniskuzZ: How this is solved in Hive?

i think it's done in `Hive: Support timestamp with local zone in Hive3 (#1897)`

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java (hunk 3 above)
Comment (pvary) [reply orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java:341]: Why is this logic is not part of the `getPromotedType` method?

**Comment by pvary** [comment]
> > @deniskuzZ: How this is solved in Hive?
> 
> i think it's done in `Hive: Support timestamp with local zone in Hive3 (#1897)`

#1897 does not have any ORC specific part. I still miss some part of the puzzle...

@zzzzming95: What happens if someone writes to this ORC table some new rows after migration?

**Comment by tanvn** [comment]
@zzzzming95
Could you take another look at the comments when you have time? 🙇

File: orc/src/test/java/org/apache/iceberg/orc/TestExpressionToSearchArgument.java [hunk 4]
```diff
@@ -332,6 +334,7 @@ public void testEvolvedSchema() {
 
   @Test
   public void testOriginalSchemaNameMapping() {
+    Configuration config = new Configuration();
```
Comment (pvary) [inline orc/src/test/java/org/apache/iceberg/orc/TestExpressionToSearchArgument.java:337]: Is this used anywhere?

File: orc/src/test/java/org/apache/iceberg/orc/TestExpressionToSearchArgument.java [hunk 5]
```diff
@@ -358,6 +363,7 @@ public void testOriginalSchemaNameMapping() {
 
   @Test
   public void testModifiedSimpleSchemaNameMapping() {
+    Configuration config = new Configuration();
```
Comment (pvary) [inline orc/src/test/java/org/apache/iceberg/orc/TestExpressionToSearchArgument.java:366]: Is this used anywhere?

File: orc/src/test/java/org/apache/iceberg/orc/TestExpressionToSearchArgument.java [hunk 6]
```diff
@@ -396,6 +402,7 @@ public void testModifiedSimpleSchemaNameMapping() {
 
   @Test
   public void testModifiedComplexSchemaNameMapping() {
+    Configuration config = new Configuration();
```
Comment (pvary) [inline orc/src/test/java/org/apache/iceberg/orc/TestExpressionToSearchArgument.java:405]: Same as above

**Comment by pvary** [comment]
> > @zzzzming95: What happens if someone writes to this ORC table some new rows after migration?
> > @pvary
> 
> Sorry I've had a recent job change, so I don't have an environment to test this case right now. But I understand that there shouldn't be a problem, because 'shouldAdjustToUTC()' will find the correct timestamp type.

Could we add a test with actual data files with the new, old, mixed format?

**Comment by tanvn** [comment]
@zzzzming95 
Sorry for bothering you again, may I ask if you could dedicate some effort to this?

**Comment by pravin1406** [comment]
@zzzzming95  I was facing this original issue and went about solving in the way you did, but it didn't work out for me,it was giving incorrect timestamp (future +5:30 (IST)).
I see my solution is almost exactly as yours. Can you confirm if you tested your PR ? Did it work ?

**Pull request closed without merging**